### DISCUSSION TITLE
How I did the "Two-class microscopy segmentation with 2D U-Net" tutorial on MacOS (M1, Sonoma 14.1.1)

### DISCUSSION THREAD

**Discussion by ValeriePineauNoel** [ivadomed/ivadomed#1314] [General]
Hi! I'm very new to all of this training-UNet-ivadomed-AxonDeepSeg thing, so I am speaking as a (very) beginner by saying that a lot of information in the tutorial is missing to make it user-friendly and accessible for non-experts. It took me a few hours to figure stuff out, so here is how I, as a non-expert, did the tutorial. I don't understand it all, so please don't hesitate to add to this discussion. 

------------------
We all start here right : [Two-class microscopy segmentation with 2D U-Net tutorial](https://ivadomed.org/tutorials/two_class_microscopy_seg_2d_unet.html). The first step is to download the dataset. **It will not work if you haven't installed and activated an ivadomed virtual environment**. Start with the installation of ivadomed: [Installation for ivadomed](https://ivadomed.org/installation.html). 
_Just for the sake of all the very beginners of this world : before installing ivadomed, go in the folder where you want to install your virtual environment with the terminal command `cd`. I'll leave this here just in case : [Mac terminal commands](https://www.techrepublic.com/article/16-terminal-commands-every-user-should-know/)._
By following these installation steps, you will see in your terminal that your hostname changes from `(base) ...` to `(ivadomed_env) (base) ...`. This will create a folder `ivadomed_env` in the directory you're in. Let's call this location `<my_current_directory>` for the rest of this discussion. 

Go back to the tutorial page and now you will be able to download the dataset `ivadomed_download_data -d data_axondeepseg_sem`. You will know it worked if the last line is `Folder Created: <my_current_directory>/data_axondeepseg_sem`. It worked instantly for me. 

The next step of the tutorial is what took me a while to figure out. You need to copy the file `config_microscopy.json`, but where is this file exactly? For me, it was under `/Users/my_user/<my_current_directory>/ivadomed_env/lib/python3.9/site-packages/ivadomed/config/config_microscopy.json`. I copied the file `config_microscopy.json` in my folder `<my_current_directory>`. 

Third step is to open the file `config_microscopy.json` in a text editor. This shouldn't be a problem. Just make sure that all the parameters are set as they say in the tutorial. For me, I had to change `"num_epochs": 200` to `"num_epochs": 50`, but the rest was the same. 

Okay, we are ready to train the model. I simply copy-pasted the command `ivadomed -c config_microscopy.json` in the terminal, without changing my current location (I'm still in `<my_current_directory>`). I have my first error `ValueError: ivadomed/config/config_bids.json is not a valid path.`. In my Finder, I can find the file `config_bids.json` easily under `<my_current_directory>/ivadomed_env/lib/python3.9/site-packages/ivadomed/config/config_bids.json`. It's basically at the same location as the file `config_microscopy.json` from before. I solved this error by changing `"loader_parameters":"bids_config"` in the `config_microscopy.json` file to the path of the file `config_bids.json`. 
_You can get the path in your Finder with a right-click on the file, then keep the option key pressed. There will be an option "Copy config_bids.json as Pathname"._ 

Let's try again. I ran `ivadomed -c config_microscopy.json` in the terminal. I don't have this `config_bids.json` error anymore but now I have a new error : `ValueError: operands could not be broadcast together with shapes (744,1154,2) (4,)`. This error is discussed in #1312. I see that the problem was solved and the solution was merged to main. I figured I might have an old version of ivadomed so I upgraded pip and ivadomed with `pip install --upgrade pip` and `pip install --upgrade ivadomed`. It didn't work. (@mathieuboudreau, @kanishk16 ? :) ) 
```
[...]
2024-03-08 12:35:28.609 | INFO     | ivadomed.loader.mri2d_segmentation_dataset:determine_cache_need:410 - Estimated 2D dataset size is 4.82853502035141e-05 GB, which is smaller than 3.013151168823242 GB. File cache will not be used
Traceback (most recent call last):
  File "/Users/valeriepineaunoel/opt/anaconda3/bin/ivadomed", line 8, in <module>
    sys.exit(run_main())
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/main.py", line 623, in run_main
    run_command(context=context,
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/main.py", line 440, in run_command
    ds_train = get_dataset(bids_df, loader_params, train_lst, transform_train_params, cuda_available, device,
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/main.py", line 129, in get_dataset
    ds = imed_loader.load_dataset(bids_df, **{**loader_params, **{'data_list': data_lst,
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/loader/loader.py", line 115, in load_dataset
    dataset.load_filenames()
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/loader/mri2d_segmentation_dataset.py", line 127, in load_filenames
    roi_pair = SegmentationPair(input_filenames,
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/loader/segmentation_pair.py", line 59, in __init__
    input_img = self.read_file(input_file)
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/loader/segmentation_pair.py", line 297, in read_file
    img = self.convert_file_to_nifti(filename, extension, is_gt)
  File "/Users/valeriepineaunoel/opt/anaconda3/lib/python3.9/site-packages/ivadomed/loader/segmentation_pair.py", line 360, in convert_file_to_nifti
    _img = np.sum(_img * (.299, .587, .114, 0), axis=-1)
ValueError: operands could not be broadcast together with shapes (744,1154,2) (4,)
```

------------------
This is where I'm at at the moment! I'll keep you informed of my progress as it goes. Please, comment, suggest stuff, help as much as you can. 
Peace,

**Comment by mathieuboudreau** [comment]
Hi @ValeriePineauNoel ! Since your issue touches closely on AxonDeepSeg, we're thinking maybe you post in our repo: https://github.com/axondeepseg/axondeepseg

Specifically, we've been working towards discontinuing our IVADOMED-based models, and @hermancollin has been working on more generalist models which should not need to be retrained to be compatible with different datasets. He can tell you more about how this model might perform with the CARS dataset "out-of-the-box" and the approximate timeline of when it should be integrated into the main ADS software.

**Reply by ValeriePineauNoel** [reply]
Thank you for the answer! I would be interested to discuss more with @hermancollin about this but, in the meantime, I'm trying to make this tutorial work to better understand how to train ADS with my own dataset. I'm stuck at this error `ValueError: operands could not be broadcast together with shapes (744,1154,2) (4,)`. You seem to have solved the problem in #1312, but it doesn't work for me.

**Reply by mathieuboudreau** [reply]
@ValeriePineauNoel Gotcha! Typically that error means that you had a matrix/image in one set of dimensions (just for simplicity's sake, I'm going to say 2x2, so 4 voxels in total), but you're trying to assign those value in a different matrix that doesn't have the same number of pixels/values (eg, you can't assign each value in a 2x2 image into a 1x2 image, for example, since one has 4 pixels and the other one only has 2). This happens sometimes in ADS because we're resizing/resampling/reassigning arrays at different points. The place where you're getting the error:

>_img = np.sum(_img * (.299, .587, .114, 0), axis=-1)

deals with converting a colour image (which has the dimensions of the image, time 3 (if a red-green-blue image) or times 4 (if it's red-green-blue + a transparency dimension). So, the error might be due to you having an image image format that this line doesn't expect, or if we had a bug for this case. Would you mind sharing the image (if you know it's a specific one) that you're getting this error for? If you got this while training a full image set, you should get the same error if you were trying to segment the image using any of our models since that conversion happens while loading the image in AxonDeepSeg

**Reply by ValeriePineauNoel** [reply]
Ok I get it! I'm simply following the tutorial here [Two-class microscopy segmentation with 2D U-Net](https://ivadomed.org/tutorials/two_class_microscopy_seg_2d_unet.html). I'm using the dataset downloaded with this line `ivadomed_download_data -d data_axondeepseg_sem`.

**Reply by mathieuboudreau** [reply]
Ok let me try and reproduce and will get back to you!

**Reply by kanishk16** [reply]
Hi @ValeriePineauNoel, @mathieuboudreau beautifully summarized the reason for the error, but I tried reproducing the error, and it seems like we didn't release a new version of ivadomed after fixing the issue. I'd suggest you install ivadomed from the master branch using `pip install git+https://github.com/ivadomed/ivadomed.git@master` till we cut a fresh release.

**Reply by ValeriePineauNoel** [reply]
Hi @kanishk16! I installed ivadomed from the master branch and reran `ivadomed -c config_microscopy.json` and had the same error (`ValueError: operands could not be broadcast together with shapes (744,1154,2) (4,).`) Here are the last lines after running `pip install git+https://github.com/ivadomed/ivadomed.git@master`:
```
...
Requirement already satisfied: docopt>=0.6.2 in ./ivadomed_env/lib/python3.9/site-packages (from num2words->pybids<0.15.6,>=0.14.0->ivadomed==2.9.9) (0.6.2)
Requirement already satisfied: mpmath>=0.19 in ./ivadomed_env/lib/python3.9/site-packages (from sympy->onnxruntime!=1.16.0,>=1.7.0->ivadomed==2.9.9) (1.3.0)
Requirement already satisfied: colorama<0.5.0,>=0.4.3 in ./ivadomed_env/lib/python3.9/site-packages (from typer[all]->torchio>=0.18.68->ivadomed==2.9.9) (0.4.6)
Requirement already satisfied: shellingham<2.0.0,>=1.3.0 in ./ivadomed_env/lib/python3.9/site-packages (from typer[all]->torchio>=0.18.68->ivadomed==2.9.9) (1.5.4)
Requirement already satisfied: rich<14.0.0,>=10.11.0 in ./ivadomed_env/lib/python3.9/site-packages (from typer[all]->torchio>=0.18.68->ivadomed==2.9.9) (13.7.1)
Requirement already satisfied: smmap<6,>=3.0.1 in ./ivadomed_env/lib/python3.9/site-packages (from gitdb<5,>=4.0.1->GitPython!=3.1.29,>=1.0.0->wandb>=0.12.11->ivadomed==2.9.9) (5.0.1)
Requirement already satisfied: markdown-it-py>=2.2.0 in ./ivadomed_env/lib/python3.9/site-packages (from rich<14.0.0,>=10.11.0->typer[all]->torchio>=0.18.68->ivadomed==2.9.9) (3.0.0)
Requirement already satisfied: pygments<3.0.0,>=2.13.0 in ./ivadomed_env/lib/python3.9/site-packages (from rich<14.0.0,>=10.11.0->typer[all]->torchio>=0.18.68->ivadomed==2.9.9) (2.17.2)
Requirement already satisfied: mdurl~=0.1 in ./ivadomed_env/lib/python3.9/site-packages (from markdown-it-py>=2.2.0->rich<14.0.0,>=10.11.0->typer[all]->torchio>=0.18.68->ivadomed==2.9.9) (0.1.2)
```
I'm not sure it worked well because it seems everything was satisfied so I'm not sure it replaced what was already there. Could you guide me on what's the best way to replace ivadomed in my folder `ivadomed_venv` without breaking everything? 
I tried to go in the folder `<my_current_location>/ivadomed_env/lib/python3.9/site-packages/` and delete ivadomed directly and rerun `pip install git+https://github.com/ivadomed/ivadomed.git@master` and running `ivadomed -c config_microscopy.json` but it still gave me the same `ValueError: operands...` error.

**Reply by kanishk16** [reply]
I'm sharing a [Google Colab Notebook](https://colab.research.google.com/drive/1F7jaRWiqWbR9nzPD0-Q84txcjTn6wjYG?usp=sharing). Now what is a Google Colab Notebook? As per the [docs](https://colab.google/), "_Colab is a hosted Jupyter Notebook service that requires no setup to use and provides free access to computing resources, including GPUs and TPUs. Colab is especially well suited to machine learning, data science, and education_." So, while experimenting it's easier to just spin up a Google Colab Notebook, without having to worry about issues related ranging from breaking your virtual environment to having access to GPUs.

**Reply by kanishk16** [reply]
@ValeriePineauNoel as far as local setup is concerned, I'd recommend creating a fresh virtual environment and then installing ivadomed using `pip install git+https://github.com/ivadomed/ivadomed.git@master`.

**Reply by ValeriePineauNoel** [reply]
I just redid all the steps on my computer by following your GoogleColab and I works until the last step where it stops with the same error `ValueError: operands could not be broadcast together with shapes (744,1154,2) (4,).`. I'm not sure why it is working in your Colab but not on my own end.
